File: onvif-gui/src/camerapanel.cpp
/*******************************************************************************
* camerapanel.cpp
*
* Copyright (c) 2020 Stephen Rhodes
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License along
* with this program; if not, write to the Free Software Foundation, Inc.,
* 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*
*******************************************************************************/

#include <sstream>
#include <algorithm>
#include "camerapanel.h"
#include "mainwindow.h"
#include <QGridLayout>
#include <QThreadPool>
#include <QMessageBox>
#include <QDateTime>

CameraPanel::CameraPanel(QMainWindow *parent)
{
    mainWindow = parent;

    videoTab = new VideoTab(this);
    imageTab = new ImageTab(this);
    networkTab = new NetworkTab(this);
    ptzTab = new PTZTab(this);
    adminTab = new AdminTab(this);

    tabWidget = new QTabWidget();
    tabWidget->addTab(videoTab, "Video");
    tabWidget->addTab(imageTab, "Image");
    tabWidget->addTab(networkTab, "Network");
    tabWidget->addTab(ptzTab, "PTZ");
    tabWidget->addTab(adminTab, "Admin");

    btnApply = new QPushButton(this);
    btnApply->setStyleSheet(MW->getButtonStyle("apply"));
    connect(btnApply, SIGNAL(clicked()), this, SLOT(btnApplyClicked()));

    btnDiscover = new QPushButton(this);
    btnDiscover->setStyleSheet(MW->getButtonStyle("discover"));
    connect(btnDiscover, SIGNAL(clicked()), this, SLOT(btnDiscoverClicked()));

    btnPlay = new QPushButton(this);
    btnPlay->setStyleSheet(MW->getButtonStyle("play"));
    connect(btnPlay, SIGNAL(clicked()), this, SLOT(btnPlayClicked()));

    btnRecord = new QPushButton(this);
    btnRecord->setStyleSheet(MW->getButtonStyle("record"));
    connect(btnRecord, SIGNAL(clicked()), this, SLOT(btnRecordClicked()));

    sldVolume = new QSlider(Qt::Horizontal, this);
    connect(sldVolume, SIGNAL(valueChanged(int)), MW, SLOT(setPlayerVolume(int)));

    btnMute = new QPushButton();
    connect(btnMute, SIGNAL(clicked()), MW, SLOT(togglePlayerMute()));

    QWidget *controlPanel = new QWidget(this);
    QGridLayout* controlLayout = new QGridLayout(controlPanel);
    controlLayout->addWidget(btnPlay,      0, 0, 1, 1);
    controlLayout->addWidget(btnRecord,    0, 1, 1, 1);
    controlLayout->addWidget(btnMute,      0, 2, 1, 1);
    controlLayout->addWidget(sldVolume,    0, 3, 1, 1);
    controlLayout->addWidget(btnDiscover,  0, 4, 1, 1);
    controlLayout->addWidget(btnApply,     0, 5, 1 ,1);
    controlPanel->setMaximumHeight(60);

    cameraList = new QListWidget(this);
    connect(cameraList, SIGNAL(itemDoubleClicked(QListWidgetItem*)), this, SLOT(cameraListDoubleClicked(QListWidgetItem*)));
    connect(cameraList, SIGNAL(itemClicked(QListWidgetItem*)), this, SLOT(cameraListClicked(QListWidgetItem*)));

    QGridLayout *layout = new QGridLayout(this);
    layout->addWidget(cameraList,     0, 0, 1, 1);
    layout->addWidget(tabWidget,      1, 0, 1, 1);
    layout->addWidget(controlPanel,   2, 0, 1, 1);
    layout->setRowStretch(0, 10);

    videoTab->setActive(false);
    imageTab->setActive(false);
    networkTab->setActive(false);
    ptzTab->setActive(false);
    adminTab->setActive(false);
    btnApply->setEnabled(false);

    connect(MW, SIGNAL(updateUI()), this, SLOT(onUpdateUI()));
    connect(this, SIGNAL(msg(const QString&)), mainWindow, SLOT(msg(const QString&)));

    //cameraNames = new QSettings("Onvif", "Camera Names");
    //foreach(QString key, cameraNames->allKeys()) {
        //discovery->cameraAlias.insert(key, cameraNames->value(key).toString());
    //}

    loginDlg = new LoginDialog();
    connect(this, SIGNAL(showLogin()), this, SLOT(onShowLogin()));
    connect(this, SIGNAL(initTabs()), this, SLOT(onInitTabs()));
    connect(this, SIGNAL(enableDiscoverButton()), this, SLOT(onEnableDiscoverButton()));

    disableToolTips(MW->settingsPanel->hideToolTips->isChecked());

    if (MW->settingsPanel->autoDiscovery->isChecked())
        btnDiscoverClicked();
}

CameraPanel::~CameraPanel()
{

}

void CameraPanel::setPlayButton()
{
    if (MW->player) btnPlay->setStyleSheet(MW->getButtonStyle("stop"));
    else btnPlay->setStyleSheet(MW->getButtonStyle("play"));
}

void CameraPanel::setMuteButton(bool mute)
{
    if (mute) btnMute->setStyleSheet(MW->getButtonStyle("mute"));
    else btnMute->setStyleSheet(MW->getButtonStyle("audio"));
}

void CameraPanel::setRecordButton()
{
    if (MW->player) {
        if (MW->player->isPiping()) btnRecord->setStyleSheet(MW->getButtonStyle("recording"));
        else btnRecord->setStyleSheet(MW->getButtonStyle("record"));
    }
    else {
        btnRecord->setStyleSheet(MW->getButtonStyle("record"));
    }
}

void CameraPanel::cameraListDoubleClicked(QListWidgetItem* item)
{
    if (connecting) {
        std::cout << "currently attempting to connect to " << MW->currentMedia.toLatin1().data() << " please wait" << std::endl;
    }
    else {
        currentStreamingRow = cameraList->row(item);
        currentDataRow = currentStreamingRow;
        std::cout << "currentStreamingRow: " << currentStreamingRow << std::endl;
        std::cout << "devices size: " << devices.size() << std::endl;
        onvif::Data data = devices[currentStreamingRow];
        std::cout << data->camera_name << std::endl;
        MW->currentMedia = data->camera_name;
        std::stringstream ss_uri;
        std::string uri(data->stream_uri);
        ss_uri << uri.substr(0, 7) << data->username << ":" << data->password << "@" << uri.substr(7);
        uri = ss_uri.str();
        connecting = true;
        MW->playerStart(uri.c_str());
    }
}

void CameraPanel::btnPlayClicked()
{
    std::cout << "btnPlayClicked" << std::endl;
    /*
    if (MW->player) {
        MW->playerStop();
    }
    else {
        if (cameraList->getCurrentCamera()) {
            QString name = cameraList->getCurrentCamera()->getCameraName();
            cameraListDoubleClicked();
        }
    }
    */
}

void CameraPanel::btnApplyClicked()
{
    CameraDialogTab *tab = (CameraDialogTab *)tabWidget->currentWidget();
    tab->update();
}

void CameraPanel::btnRecordClicked()
{
    std::cout << "record button clicked - use environment variable QT_FILESYSTEMMODEL_WATCH_FILES for file size updates" << std::endl;
    recording = !recording;
    QString filename = MW->filePanel->directorySetter->directory;
    if (MW->settingsPanel->generateFilename->isChecked()) 
        filename.append("/").append(QDateTime::currentDateTime().toString("yyyyMMddhhmmss")).append(".mp4");
    else 
        filename.append("/").append("out.mp4");

    if (MW->player) {
        MW->player->togglePiping(filename.toLatin1().data());
    }
    setRecordButton();
}

void CameraPanel::disableToolTips(bool arg)
{
    if (!arg)
    {
        btnApply->setToolTip("Apply");
        btnDiscover->setToolTip("Discover");
        btnPlay->setToolTip("Play");
        btnRecord->setToolTip("Record");
        btnMute->setToolTip("Mute");
    }
    else {
        btnApply->setToolTip("");
        btnDiscover->setToolTip("");
        btnPlay->setToolTip("");
        btnRecord->setToolTip("");
        btnMute->setToolTip("");
    }
}

/*
void CameraPanel::fillData()
{
    videoTab->clear();
    imageTab->clear();
    networkTab->clear();
    adminTab->clear();
    videoTab->setActive(false);
    imageTab->setActive(false);
    networkTab->setActive(false);
    adminTab->setActive(false);
    btnApply->setEnabled(false);
    //QThreadPool::globalInstance()->tryStart(filler);
}
*/

void CameraPanel::showData()
{
    videoTab->initialize();
    imageTab->initialize();
    networkTab->initialize();
    adminTab->initialize();

    videoTab->setActive(true);
    imageTab->setActive(true);
    networkTab->setActive(true);
    adminTab->setActive(true);
    if (currentDataRow > -1) ptzTab->setActive(hasPTZ(devices[currentDataRow]));
    btnApply->setEnabled(false);   
}

void CameraPanel::refreshList()
{
    //cameraList->refresh();
}

void CameraPanel::onUpdateUI()
{
    setPlayButton();
}

void CameraPanel::btnDiscoverClicked()
{
    std::cout << "start discover" << std::endl;
    onvif::Manager onvifBoss;
    onvifBoss.startDiscover([&]() { discoverFinished(); },
                            [&](onvif::Data& data) { return getCredential(data); },
                            [&](onvif::Data& data) { return getData(data); }
                            );
    btnDiscover->setEnabled(false);

}

void CameraPanel::getData(onvif::Data& onvif_data)
{
    std::cout << "get data" << std::endl;
    if (std::find(devices.begin(), devices.end(), onvif_data) == devices.end()) {
        onvif_data->logged_in = true;
        devices.push_back(onvif_data);
        cameraList->addItem(onvif_data->camera_name);
    }
}

void CameraPanel::onShowLogin()
{
    loginDlg->setStyleSheet(MW->style);
    if (!loginDlg->exec())
        loginDlg->cancelled = true;
    loginDlg->active = false;
}

bool CameraPanel::getCredential(onvif::Data& onvif_data)
{
    std::cout << "get credential" << std::endl;
    if (onvif_data->logged_in) return false;

    QString username = MW->settingsPanel->commonUsername->text();
    QString password = MW->settingsPanel->commonPassword->text();

    if (username.length() == 0 || last_data == onvif_data) 
    {
        loginDlg->active = true;
        loginDlg->cameraIP->setText(onvif_data->host);
        loginDlg->cameraName->setText(onvif_data->camera_name);
        emit showLogin();

        while (loginDlg->active)
            std::this_thread::sleep_for(std::chrono::milliseconds(1));  

        if (loginDlg->cancelled) return false;

        username = loginDlg->username->text();
        password = loginDlg->password->text();
    }

    strncpy(onvif_data->username, username.toLatin1(), username.length());
    strncpy(onvif_data->password, password.toLatin1(), password.length());

    last_data = onvif_data;
    
    return true;
}

void CameraPanel::onEnableDiscoverButton()
{
    btnDiscover->setEnabled(true);
}

void CameraPanel::discoverFinished()
{
    emit enableDiscoverButton();
}

void CameraPanel::cameraListClicked(QListWidgetItem* item)
{
    std::cout << "cameraListClicked" << std::endl;
    currentDataRow = cameraList->row(item);
    if (!devices[currentDataRow]->filled) {
        onvif::Manager onvifBoss;
        onvifBoss.startFill(devices[currentDataRow], 
                        [&](const onvif::Data& onvif_data) { fillData(onvif_data); });

        cameraList->setEnabled(false);
        videoTab->setActive(false);
        imageTab->setActive(false);
        networkTab->setActive(false);
        ptzTab->setActive(false);
        adminTab->setActive(false);
    }
    else {
        showData();
    }
}

void CameraPanel::onInitTabs()
{
    showData();
    cameraList->setEnabled(true);
}

void CameraPanel::fillData(const onvif::Data& onvif_data)
{
    std::cout << "CameraPanel::fillData" << std::endl;
    devices[currentDataRow] = onvif_data;
    devices[currentDataRow]->filled = true;
    emit initTabs();
}

File: python/camerapanel.py
import os
import sys
import numpy as np
import cv2
from time import sleep
from PyQt6.QtWidgets import QDialogButtonBox, QLineEdit, QPushButton, \
QGridLayout, QWidget, QSlider, QLabel, QMessageBox, QListWidget, \
QTabWidget
from PyQt6.QtGui import QIcon
from PyQt6.QtCore import Qt, pyqtSignal, QObject, QSettings
from logindialog import LoginDialog
from videotab import VideoTab
from imagetab import ImageTab
from networktab import NetworkTab
from ptztab import PTZTab
from admintab import AdminTab

sys.path.append("../build/libonvif")
import onvif

class CameraPanelSignals(QObject):
    fill = pyqtSignal(onvif.Data)
    login = pyqtSignal(onvif.Data)


class CameraList(QListWidget):
    def __init__(self):
        super().__init__()

    def keyPressEvent(self, event):
        if event.key() == Qt.Key.Key_Return:
            self.itemDoubleClicked.emit(self.currentItem())
        return super().keyPressEvent(event)
    
class CameraPanel(QWidget):
    def __init__(self, mw):
        super().__init__()
        self.devices = []
        self.mw = mw
        self.dlgLogin = LoginDialog(self)
        self.settings = QSettings("onvif", "alias")
        self.icnDiscover = QIcon('image:discover.png')
        self.icnApply = QIcon("image:apply.png")
        self.icnMute = QIcon('image:mute.png')
        self.icnAudio = QIcon('image:audio.png')
        self.icnRecord = QIcon("image:record.png")
        self.icnRecording = QIcon("image:recording.png")

        self.boss = onvif.Manager()
        self.boss.discovered = lambda : self.discovered()
        self.boss.getCredential = lambda D : self.getCredential(D)
        self.boss.getData = lambda D : self.getData(D)
        self.boss.filled = lambda D : self.filled(D)

        self.btnRecord = QPushButton()
        self.btnRecord.setIcon(self.icnRecord)
        self.btnRecord.setToolTip("Record")
        self.btnRecord.setToolTipDuration(2000)
        self.btnRecord.setMinimumWidth(self.icnRecord.availableSizes()[0].width() * 2)
        self.btnRecord.setFocusPolicy(Qt.FocusPolicy.NoFocus)
        self.btnRecord.clicked.connect(self.btnRecordClicked)

        self.btnMute = QPushButton()
        if self.mw.mute:
            self.btnMute.setIcon(self.icnMute)
        else:
            self.btnMute.setIcon(self.icnAudio)
        self.btnMute.setToolTip("Mute")
        self.btnMute.setToolTipDuration(2000)
        self.btnMute.setMinimumWidth(self.icnMute.availableSizes()[0].width() * 2)
        self.btnMute.setFocusPolicy(Qt.FocusPolicy.NoFocus)
        self.btnMute.clicked.connect(self.btnMuteClicked)

        self.sldVolume = QSlider(Qt.Orientation.Horizontal)
        self.sldVolume.setValue(self.mw.volume)
        self.sldVolume.setFocusPolicy(Qt.FocusPolicy.NoFocus)
        self.sldVolume.valueChanged.connect(self.sldVolumeChanged)

        self.btnDiscover = QPushButton()
        self.btnDiscover.setIcon(self.icnDiscover)
        self.btnDiscover.setToolTip("Discover")
        self.btnDiscover.setToolTipDuration(2000)
        self.btnDiscover.setMinimumWidth(self.icnDiscover.availableSizes()[0].width() * 2)
        self.btnDiscover.setFocusPolicy(Qt.FocusPolicy.NoFocus)
        self.btnDiscover.clicked.connect(self.btnDiscoverClicked)

        self.btnApply = QPushButton()
        self.btnApply.setIcon(self.icnApply)
        self.btnApply.setToolTip("Apply")
        self.btnApply.setToolTipDuration(2000)
        self.btnApply.setMinimumWidth(self.icnApply.availableSizes()[0].width() * 2)
        self.btnApply.setFocusPolicy(Qt.FocusPolicy.NoFocus)
        self.btnApply.clicked.connect(self.btnApplyClicked)
        self.btnApply.setEnabled(False)
        
        self.lstCamera = CameraList()
        self.lstCamera.currentRowChanged.connect(self.onCurrentRowChanged)
        self.lstCamera.itemDoubleClicked.connect(self.onItemDoubleClicked)

        self.tabOnvif = QTabWidget()
        self.tabVideo = VideoTab(self)
        self.tabImage = ImageTab(self)
        self.tabNetwork = NetworkTab(self)
        self.ptzTab = PTZTab(self)
        self.adminTab = AdminTab(self)
        self.tabOnvif.addTab(self.tabVideo, "Video")
        self.tabOnvif.addTab(self.tabImage, "Image")
        self.tabOnvif.addTab(self.tabNetwork, "Network")
        self.tabOnvif.addTab(self.ptzTab, "PTZ")
        self.tabOnvif.addTab(self.adminTab, "Admin")

        self.signals = CameraPanelSignals()
        self.signals.fill.connect(self.tabVideo.fill)
        self.signals.fill.connect(self.tabImage.fill)
        self.signals.fill.connect(self.tabNetwork.fill)
        self.signals.fill.connect(self.ptzTab.fill)
        self.signals.fill.connect(self.adminTab.fill)
        self.signals.login.connect(self.onShowLogin)

        lytMain = QGridLayout(self)
        lytMain.addWidget(self.lstCamera,   0, 0, 1, 5)
        lytMain.addWidget(self.tabOnvif,    1, 0, 1, 5)
        lytMain.addWidget(self.btnRecord,   2, 0, 1, 1, Qt.AlignmentFlag.AlignCenter)
        lytMain.addWidget(self.btnMute,     2, 1, 1, 1, Qt.AlignmentFlag.AlignCenter)
        lytMain.addWidget(self.sldVolume,   2, 2, 1, 1)
        lytMain.addWidget(self.btnDiscover, 2, 3, 1, 1, Qt.AlignmentFlag.AlignCenter)
        lytMain.addWidget(self.btnApply,    2, 4, 1, 1, Qt.AlignmentFlag.AlignCenter)
        lytMain.setRowStretch(0, 10)

    def btnDiscoverClicked(self):
        self.boss.startPyDiscover()
        self.btnDiscover.setEnabled(False)

    def filled(self, onvif_data):
        if len(onvif_data.last_error()) > 0:
            print("ERROR:", onvif_data.last_error())
        self.devices[self.lstCamera.currentRow()] = onvif_data
        self.signals.fill.emit(onvif_data)
        self.btnApply.setEnabled(False)
        if not self.mw.connecting:
            self.setEnabled(True)
            self.lstCamera.setFocus()

    def discovered(self):
        self.btnDiscover.setEnabled(True)

    def getCredential(self, onvif_data):
        for d in self.devices:
            if d == onvif_data:
                onvif_data.cancelled = True
                return onvif_data
            
        if len(self.mw.settingsPanel.txtPassword.text()) > 0 and \
                len(onvif_data.last_error()) == 0:
            onvif_data.setUsername(self.mw.settingsPanel.txtUsername.text())
            onvif_data.setPassword(self.mw.settingsPanel.txtPassword.text())
        else:
            onvif_data.clearLastError()
            self.dlgLogin.active = True
            self.signals.login.emit(onvif_data)
            while self.dlgLogin.active:
                sleep(0.01)

        return onvif_data
    
    def onShowLogin(self, onvif_data):
        self.dlgLogin.exec(onvif_data)
    
    def getData(self, onvif_data):
        onvif_data.alias = self.settings.value(onvif_data.serial_number(), onvif_data.camera_name())
        self.devices.append(onvif_data)
        self.lstCamera.addItem(onvif_data.alias)

    def onCurrentRowChanged(self, row):
        onvif_data = self.devices[self.lstCamera.currentRow()]
        if onvif_data.filled:
            self.setTabsEnabled(False)
            self.signals.fill.emit(onvif_data)
        else:
            self.boss.onvif_data = onvif_data
            self.setTabsEnabled(False)
            self.boss.startPyFill()

    def onItemDoubleClicked(self, item):
        onvif_data = self.devices[self.lstCamera.currentRow()]
        uri = onvif_data.stream_uri()[0 : 7] + onvif_data.username() + ":" \
            + onvif_data.password() + "@" + onvif_data.stream_uri()[7:]
        self.mw.connecting = True
        self.mw.playMedia(uri)

    def setTabsEnabled(self, enabled):
        self.tabVideo.setEnabled(enabled)
        self.tabImage.setEnabled(enabled)
        self.tabNetwork.setEnabled(enabled)
        self.ptzTab.setEnabled(enabled)
        self.adminTab.setEnabled(enabled)

    def btnApplyClicked(self):
        onvif_data = self.devices[self.lstCamera.currentRow()]
        self.tabVideo.update(onvif_data)
        self.tabImage.update(onvif_data)
        self.tabNetwork.update(onvif_data)
        self.adminTab.update(onvif_data)
        self.setEnabled(False)

    def onEdit(self):
        onvif_data = self.devices[self.lstCamera.currentRow()]
        if self.tabVideo.edited(onvif_data) or \
                self.tabImage.edited(onvif_data) or \
                self.tabNetwork.edited(onvif_data) or \
                self.adminTab.edited(onvif_data):
            self.btnApply.setEnabled(True)
        else:
            self.btnApply.setEnabled(False)

    def onMediaStarted(self, n):
        self.setEnabled(True)
        if self.mw.tab.currentIndex() == 0:
            self.lstCamera.setFocus()
            self.mw.setWindowTitle(self.lstCamera.currentItem().text())

    def sldVolumeChanged(self, value):
        self.mw.filePanel.control.sldVolume.setValue(value)
        self.mw.setVolume(value)

    def setBtnMute(self):
        if self.mw.mute:
            self.btnMute.setIcon(self.icnMute)
        else:
            self.btnMute.setIcon(self.icnAudio)
       
    def btnMuteClicked(self):
        self.mw.toggleMute()
        self.setBtnMute()
        self.mw.filePanel.control.setBtnMute()

    def btnRecordClicked(self):
        print("btnRecordClicked")